include/quic/context.hpp

```diff
@@ -12,73 +12,52 @@
-
-        config_t() = default;
+        // datagram support
+        bool datagram_support = false;
+        // datagram splitting support
+        bool split_packet = false;
+        // splitting policy
+        Splitting policy = Splitting::NONE;
+
+        user_config() = default;
     };
 
-    struct ContextBase
+    struct IOContext
     {
       public:
+        Direction dir;
         std::shared_ptr<TLSCreds> tls_creds;
-        stream_data_callback_t stream_data_cb;
-        stream_open_callback_t stream_open_cb;
-        stream_close_callback_t stream_close_cb;
-        config_t config{};
-
-        // TODO: I think we can move the handle_opt calls here
-
-        virtual ~ContextBase() = default;
-    };
+        stream_data_callback stream_data_cb;
+        stream_open_callback stream_open_cb;
+        stream_close_callback stream_close_cb;
+        user_config config{};
 
-    struct OutboundContext : public ContextBase
-    {
         template <typename... Opt>
-        OutboundContext(Opt&&... opts)
+        IOContext(Direction d, Opt&&... opts) : dir{d}
         {
-            log::trace(log_cat, "Making outbound session context...");
+            log::trace(log_cat, "Making IO session context");
             // parse all options
-            ((void)handle_outbound_opt(std::forward<Opt>(opts)), ...);
-
-            log::debug(log_cat, "Outbound session context created successfully");
-        }
-
-      private:
-        void handle_outbound_opt(std::shared_ptr<TLSCreds> tls);
-        void handle_outbound_opt(opt::max_streams ms);
-        void handle_outbound_opt(stream_data_callback_t func);
-        void handle_outbound_opt(stream_open_callback_t func);
-        void handle_outbound_opt(stream_close_callback_t func);
-    };
+            ((void)handle_ioctx_opt(std::forward<Opt>(opts)), ...);
 
-    struct InboundContext : public ContextBase
-    {
-        template <typename... Opt>
-        InboundContext(Opt&&... opts)
-        {
-            log::trace(log_cat, "Making inbound session context...");
-            // parse all options
-            ((void)handle_inbound_opt(std::forward<Opt>(opts)), ...);
+            if (tls_creds == nullptr)
+                throw std::runtime_error{"Session IOContext requires some form of TLS credentials to operate"};
 
-            log::debug(log_cat, "Inbound session context successfully created");
+            log::debug(
+                    log_cat, "{} IO context created successfully", (dir == Direction::OUTBOUND) ? "Outbound"s : "Inbound"s);
         }
 
+        ~IOContext() = default;
+
       private:
-        void handle_inbound_opt(std::shared_ptr<TLSCreds> tls);
-        void handle_inbound_opt(opt::max_streams ms);
-        void handle_inbound_opt(stream_data_callback_t func);
-        void handle_inbound_opt(stream_open_callback_t func);
-        void handle_inbound_opt(stream_close_callback_t func);
+        void handle_ioctx_opt(std::shared_ptr<TLSCreds> tls);
```

Comment:
I think this one should be explicit from `Endpoint` (as you did with `Direction`) and the Endpoint should require a non-null `TLSCreds` on construction.  i.e. an Endpoint should have one `TLSCreds` shared among all connections, inbound or outbound.

Reply:
makes sense to me :+1: